R/data_separate.R

```diff
@@ -0,0 +1,403 @@
+  guess_columns <- match.arg(guess_columns, choices = c("min", "max", "mode"))
+
+  # make sure we have valid options for fill and extra
+  fill <- match.arg(fill, choices = c("left", "right", "value_left", "value_right"))
+  extra <- match.arg(extra, choices = c("drop_left", "drop_right", "merge_left", "merge_right"))
+
+  # evaluate select/exclude, may be select-helpers
+  select <- .select_nse(select,
+    data,
+    exclude,
+    ignore_case,
+    regex = regex,
+    verbose = verbose
+  )
+
+  # make new_columns as list, this works with single and multiple columns
+  if (!is.null(new_columns) && !is.list(new_columns)) {
+    new_columns <- rep(list(new_columns), times = length(select))
+    # if we have multiple columns that were separated, we avoid duplicated
+    # column names of created variables by appending name of original column
+    # however, we don't have duplicated column names when we merge them together
+    # so don't create new column names when "merge_multiple" is FALSE.
+    make_unique_colnames <- length(select) > 1 && !merge_multiple
+  } else {
+    # we don't want to create own unique column names when user explicitly
+    # provided column names as a list, i.e. column names for each separated
+    # variable
+    make_unique_colnames <- FALSE
+  }
+
+  # make sure list of new column names is named
+  if (!is.null(new_columns) && is.null(names(new_columns))) {
+    names(new_columns) <- select
+  }
+
+  # iterate columns that should be split
+  split_data <- lapply(select, function(sep_column) {
+
+    # do we have known number of columns?
+    if (is.null(new_columns)) {
+      n_columns <- NULL
+    } else {
+      n_columns <- length(new_columns[[sep_column]])
+    }
+
+    # make sure we have a character that we can split
+    x <- data[[sep_column]]
+    if (!is.character(x)) {
+      x <- as.character(x)
+    }
+
+    # separate column into multiple strings
+    if (is.numeric(separator)) {
+      maxlen <- max(nchar(x), na.rm = TRUE)
+      starts <- c(0, separator)
+      ends <- c(separator - 1, maxlen)
+      separated_columns <- lapply(seq_along(starts), function(i) {
+        substr(x, starts[i], ends[i])
+      })
+      separated_columns <- as.data.frame(
+        t(do.call(cbind, separated_columns)),
+        stringsAsFactors = FALSE
+      )
+    } else {
+      separated_columns <- strsplit(x, separator)
+    }
+
+    # how many new columns do we need?
+    if (is.null(n_columns)) {
+      # lengths of all split strings
+      l <- lengths(separated_columns)
+      # but without NA values
+      l <- l[!vapply(l, function(i) all(is.na(i)), TRUE)]
+      # define number of new columns, based on user-choice
+      n_cols <- switch(
+        guess_columns,
+        "min" = min(l, na.rm = TRUE),
+        "max" = max(l, na.rm = TRUE),
+        "mode" = distribution_mode(l),
+      )
+      # tell user
+      if (verbose && !all(l == 1) && !is.numeric(separator)) {
```

Comment:
```suggestion
      if (verbose && insight::n_unique(l) != 1 && !is.numeric(separator)) {
```

Reply:
This is to avoid an unwanted message like below (all values are split in 3 so there shouldn't be a message):

``` r
library(datawizard)

# multiple columns to split
d <- data.frame(
  x = c("1.a.6", "2.b.7", "3.c.8"),
  stringsAsFactors = FALSE
)
d
#>       x
#> 1 1.a.6
#> 2 2.b.7
#> 3 3.c.8
# split two columns, default column names
data_separate(d, guess_columns = "mode")
#> Column `x` had different number of values after splitting. Variable was
#>   split into 3 columns.
#>   x_1 x_2 x_3
#> 1   1   a   6
#> 2   2   b   7
#> 3   3   c   8
```